substrate/client/src/call_executor.rs

```diff
@@ -0,0 +1,191 @@
+	fn call(&self, id: &BlockId, method: &str, call_data: &[u8]) -> Result<CallResult, error::Error>;
+
+	/// Execute a call to a contract on top of given state.
+	///
+	/// No changes are made.
+	fn call_at_state<S: state_machine::Backend>(&self, state: &S, overlay: &mut OverlayedChanges, method: &str, call_data: &[u8]) -> Result<(Vec<u8>, S::Transaction), error::Error>;
+}
+
+/// Call executor that executes methods locally, querying all required
+/// data from local backend.
+pub struct LocalCallExecutor<B, E> {
+	backend: Arc<B>,
+	executor: E,
+}
+
+/// Call executor that executes methods on remote node, querying execution proof
+/// and checking proof by re-executing locally.
+pub struct RemoteCallExecutor<B, E> {
+	backend: Arc<B>,
+	executor: E,
+	fetcher: Arc<Fetcher>,
+}
+
+impl<B, E> LocalCallExecutor<B, E> {
+	/// Creates new instance of local call executor.
+	pub fn new(backend: Arc<B>, executor: E) -> Self {
+		LocalCallExecutor { backend, executor }
+	}
+}
+
+impl<B, E> Clone for LocalCallExecutor<B, E> where E: Clone {
+	fn clone(&self) -> Self {
+		LocalCallExecutor {
+			backend: self.backend.clone(),
+			executor: self.executor.clone(),
+		}
+	}
+}
+
+impl<B, E> CallExecutor for LocalCallExecutor<B, E>
+	where
+		B: backend::LocalBackend,
+		E: CodeExecutor,
+		error::Error: From<<<B as backend::Backend>::State as StateBackend>::Error>,
+{
+	type Error = E::Error;
+
+	fn call(&self, id: &BlockId, method: &str, call_data: &[u8]) -> error::Result<CallResult> {
+		let mut changes = OverlayedChanges::default();
+		let (return_data, _) = self.call_at_state(&self.backend.state_at(*id)?, &mut changes, method, call_data)?;
+		Ok(CallResult{ return_data, changes })
+	}
+
+	fn call_at_state<S: state_machine::Backend>(&self, state: &S, changes: &mut OverlayedChanges, method: &str, call_data: &[u8]) -> error::Result<(Vec<u8>, S::Transaction)> {
+		state_machine::execute(
+			state,
+			changes,
+			&self.executor,
+			method,
+			call_data,
+		).map_err(Into::into)
+	}
+}
+
+impl<B, E> RemoteCallExecutor<B, E> {
+	/// Creates new instance of remote call executor.
+	pub fn new(backend: Arc<B>, executor: E, fetcher: Arc<Fetcher>) -> Self {
+		RemoteCallExecutor { backend, executor, fetcher }
+	}
+}
+
+impl<B, E> CallExecutor for RemoteCallExecutor<B, E>
+	where
+		B: backend::RemoteBackend,
+		E: CodeExecutor,
+		error::Error: From<<<B as backend::Backend>::State as StateBackend>::Error>,
+{
+	type Error = E::Error;
+
+	fn call(&self, id: &BlockId, method: &str, call_data: &[u8]) -> error::Result<CallResult> {
+		let block_hash = match *id {
+			BlockId::Hash(hash) => hash,
+			BlockId::Number(number) => self.backend.blockchain().hash(number)?
+				.ok_or_else(|| error::ErrorKind::UnknownBlock(BlockId::Number(number)))?,
+		};
+
+		let (remote_result, remote_proof) = self.fetcher.execution_proof(block_hash, method, call_data)?;
+
+		// code below will be replaced with proper proof check once trie-based proofs will be possible
```

Comment:
these are possible now

Reply:
yeah - better to start adding support for tries after this PR (overall structure) is approved